diff --git a/lib/tray/meson.build b/lib/tray/meson.build
index fbf2f98..1a8fc3f 100644
--- a/lib/tray/meson.build
+++ b/lib/tray/meson.build
@@ -39,29 +39,9 @@ deps = [
   dependency('gio-2.0'),
   dependency('json-glib-1.0'),
   dependency('gdk-pixbuf-2.0'),
-  dependency('gtk+-3.0'),
+  dependency('appmenu-glib-translator')
 ]
 
-dbusmenu_cflags = run_command(
-  find_program('pkg-config', required: true),
-  '--cflags', 'dbusmenu-gtk3-0.4',
-  'gobject-introspection-1.0',
-  'gobject-2.0',
-  'glib-2.0',
-  capture: true,
-  check: true,
-).stdout().strip()
-
-dbusmenu_libs = run_command(
-  find_program('pkg-config', required: true),
-  '--libs', 'dbusmenu-gtk3-0.4',
-  'gobject-introspection-1.0',
-  'gobject-2.0',
-  'glib-2.0',
-  capture: true,
-  check: true,
-).stdout().strip()
-
 sources = [config] + files('tray.vala', 'watcher.vala', 'trayItem.vala')
 
 if get_option('lib')
@@ -71,15 +51,13 @@ if get_option('lib')
     dependencies: deps,
     vala_header: meson.project_name() + '.h',
     vala_vapi: meson.project_name() + '-' + api_version + '.vapi',
-    vala_args: ['--vapi-comments', '--pkg', 'DbusmenuGtk3-0.4', '--pkg', 'Dbusmenu-0.4'],
+    vala_args: ['--vapi-comments'],
     version: meson.project_version(),
-    c_args: dbusmenu_cflags.split(' '),
-    link_args: dbusmenu_libs.split(' '),
     install: true,
     install_dir: [true, true, true]
   )
 
-  pkgs = ['--pkg', 'DbusmenuGtk3-0.4', '--pkg', 'Dbusmenu-0.4']
+  pkgs = []
   foreach dep : deps
     pkgs += ['--pkg=' + dep.name()]
   endforeach
@@ -120,31 +98,13 @@ if get_option('lib')
     requires: deps,
     install_dir: get_option('libdir') / 'pkgconfig',
   )
-  #
-  # custom_target(
-  #   typelib,
-  #   command: [
-  #     find_program('g-ir-compiler'),
-  #     '--output', '@OUTPUT@',
-  #     '--shared-library', get_option('prefix') / get_option('libdir') / '@PLAINNAME@',
-  #     meson.current_build_dir() / gir,
-  #   ],
-  #   input: lib,
-  #   output: typelib,
-  #   depends: lib,
-  #   install: true,
-  #   install_dir: get_option('libdir') / 'girepository-1.0',
-  # )
-endif
+  endif
 
 if get_option('cli')
   executable(
     meson.project_name(),
     ['cli.vala', sources],
     dependencies: deps,
-    vala_args: ['--pkg', 'DbusmenuGtk3-0.4', '--pkg', 'Dbusmenu-0.4'],
-    c_args: dbusmenu_cflags.split(' '),
-    link_args: dbusmenu_libs.split(' '),
     install: true,
   )
-endif
+endif
\ No newline at end of file
diff --git a/lib/tray/trayItem.vala b/lib/tray/trayItem.vala
index db0e6d4..d416554 100644
--- a/lib/tray/trayItem.vala
+++ b/lib/tray/trayItem.vala
@@ -1,5 +1,3 @@
-using DbusmenuGtk;
-
 namespace AstalTray {
 public struct Pixmap {
     int width;
@@ -169,6 +167,22 @@ public class TrayItem : Object {
     /** the id of the item used to uniquely identify the TrayItems by this lib.*/
     public string item_id { get; private set; }
 
+    private DBusMenu.Importer menu_importer;
+     
+    public MenuModel menu_model {
+      owned get {
+        if (menu_importer == null) return null;
+        return menu_importer.model;
+      }
+    }
+
+    public ActionGroup action_group {
+      owned get {
+        if (menu_importer == null) return null;
+        return menu_importer.action_group;
+      }
+    }
+
     public signal void changed();
     public signal void ready();
 
@@ -197,6 +211,14 @@ public class TrayItem : Object {
                 }
             });
            
+            if(proxy.Menu != null) {
+              this.menu_importer = new DBusMenu.Importer(proxy.get_name_owner(), proxy.Menu);
+              this.menu_importer.notify["model"].connect(() => {
+                notify_property("menu-model");
+                notify_property("action-group");
+              });
+            }
+
             update_gicon();
 
             ready();
@@ -216,18 +238,12 @@ public class TrayItem : Object {
 
     private void update_gicon() {
         if(icon_name != null && icon_name != "") {
-            if(icon_theme_path != null && icon_theme_path != "") {
-
-                Gtk.IconTheme icon_theme = new Gtk.IconTheme();
-                string[] paths = {icon_theme_path};
-                icon_theme.set_search_path(paths);
-
-                int size = icon_theme.get_icon_sizes(icon_name)[0];
-                Gtk.IconInfo icon_info = icon_theme.lookup_icon(
-                    icon_name, size, Gtk.IconLookupFlags.FORCE_SIZE);
-
-                if (icon_info != null)
-                    gicon = new GLib.FileIcon(GLib.File.new_for_path(icon_info.get_filename()));
+            if(GLib.FileUtils.test(icon_name, GLib.FileTest.EXISTS)) {
+                gicon = new GLib.FileIcon(GLib.File.new_for_path(icon_name));
+            }
+            else if(icon_theme_path != null && icon_theme_path != "") {
+                gicon = new GLib.FileIcon(GLib.File.new_for_path(
+                    find_icon_in_theme(icon_name, icon_theme_path)));
             } else {
                 gicon = new GLib.ThemedIcon(icon_name);
             }
@@ -312,16 +328,33 @@ public class TrayItem : Object {
         }
     }
 
-    /**
-    * creates a new Gtk Menu for this item.
-    */
-    public Gtk.Menu? create_menu() {
-        if (proxy.Menu == null)
-            return null;
+    private string? find_icon_in_theme(string icon_name, string theme_path){
+      if(icon_name == null || theme_path == null || icon_name == "" || theme_path == "") return null;
+ 
+      try {
+          Dir dir = Dir.open (theme_path, 0);
+          string? name = null;
+
+          while ((name = dir.read_name ()) != null) {
+            string path = Path.build_filename (theme_path, name);
+            
+            if (FileUtils.test (path, FileTest.IS_DIR)) {
+              string? icon = find_icon_in_theme(icon_name, path);
+              if(icon != null) return icon;
+              else continue;
+            }
+
+            int dot_index = name.last_index_of(".");
+            if (dot_index != -1)
+                name = name.substring(0, dot_index);
+            if (name == icon_name) return path;
+
+          }
+      } catch (FileError err) {
+          return null;
+      }
+      return null;
 
-        return new DbusmenuGtk.Menu(
-            proxy.get_name_owner(),
-            proxy.Menu);
     }
 
     private Gdk.Pixbuf? _get_icon_pixbuf() {
@@ -329,41 +362,7 @@ public class TrayItem : Object {
             ? proxy.AttentionIconPixmap
             : proxy.IconPixmap;
 
-
-        string icon_name = proxy.Status == Status.NEEDS_ATTENTION
-            ? proxy.AttentionIconName
-            : proxy.IconName;
-
-        Gdk.Pixbuf pixbuf = null;
-
-        if (icon_name != null && proxy.IconThemePath != null)
-            pixbuf = load_from_theme(icon_name, proxy.IconThemePath);
-
-        if (pixbuf == null)
-            pixbuf = pixmap_to_pixbuf(pixmaps);
-
-        return pixbuf;
-    }
-
-    private Gdk.Pixbuf? load_from_theme(string icon_name, string theme_path) {
-        if (theme_path == "" || theme_path == null)
-            return null;
-
-        if (icon_name == "" || icon_name == null)
-            return null;
-
-        Gtk.IconTheme icon_theme = new Gtk.IconTheme();
-        string[] paths = {theme_path};
-        icon_theme.set_search_path(paths);
-
-        int size = icon_theme.get_icon_sizes(icon_name)[0];
-        Gtk.IconInfo icon_info = icon_theme.lookup_icon(
-            icon_name, size, Gtk.IconLookupFlags.FORCE_SIZE);
-
-        if (icon_info != null)
-            return icon_info.load_icon();
-
-        return null;
+        return pixmap_to_pixbuf(pixmaps);
     }
 
     private Gdk.Pixbuf? pixmap_to_pixbuf(Pixmap[] pixmaps) {
@@ -371,6 +370,12 @@ public class TrayItem : Object {
             return null;
 
         Pixmap pixmap = pixmaps[0];
+        
+        for(int i = 0; i < pixmaps.length; i++){
+            if(pixmap.width < pixmaps[i].width)
+              pixmap = pixmaps[i];
+        };
+
         uint8[] image_data = pixmap.bytes.copy();
 
         for (int i = 0; i < pixmap.width * pixmap.height * 4; i += 4) {
@@ -417,4 +422,4 @@ public class TrayItem : Object {
             .get_root();
     }
 }
-}
+}
\ No newline at end of file
